IMOD/manpages/edgeeraser.man
.na
.nh
.TH edgeeraser 1 4.6.34 IMOD
.SH NAME
edgeeraser - erase defects along the edges of images
.SH SYNOPSIS
edgeeraser
.SH DESCRIPTION
This program erases bad portions of an image near its edges.  It is
designed to eliminate the sharp borders created when an image
digitized from film does not fill the entire frame.  The image is
tapered down to the mean intensity gradually to eliminate the strong
artifacts that would otherwise occur in a tomographic reconstruction.
.P
To use the program, first prepare an IMOD model file to indicate
which edges need to be erased.  Each edge should be specified with a
separate contour; do not go around a corner with a single contour to
specify two edges.  Draw the contour just inside the edge of the bad
region on the image.  The program will erase along that entire side
of the image, so you should generally draw a contour along the whole
length of a side.  If you do not, the program will extrapolate from
the endpoints of a contour.
.P
The inputs are:
.P
Input image file
.P
Output image file, or <Return> to place modified sections back into
the input file.  USE REPLACEMENT OPTION WITH CAUTION
.P
Model file
.P
Distance over which the image should be tapered down to the mean
intensity, or / to accept the default.  This taper will occur inside
of the drawn contour, so that the mean intensity is reached at the
position of the contour.
.SH HISTORY
.nf
	  Written by David Mastronarde  12/4/98
.fi
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/flib/model/edgeeraser.f
****    EDGEERASER.FOR
c       
c       This program erases bad portions of an image near its edges.  It is
c       designed to eliminate the sharp borders created when an image
c       digitized from film does not fill the entire frame.  The image is
c       tapered down to the mean intensity gradually to eliminate the strong
c       artifacts that would otherwise occur in a tomographic reconstruction.
c       
c       To use the program, first prepare an IMOD model file to indicate
c       which edges need to be erased.  Each edge should be specified with a
c       separate contour; do not go around a corner with a single contour to
c       specify two edges.  Draw the contour just inside the edge of the bad
c       region on the image.  The program will erase along that entire side
c       of the image, so you should generally draw a contour along the whole
c       length of a side.  If you do not, the program will extrapolate from
c       the endpoints of a contour.
c       
c       The inputs are:
c       
c       Input image file
c       
c       Output image file, or <Return> to place modified sections back into
c       the input file.  USE REPLACEMENT OPTION WITH CAUTION
c       
c       Model file
c       
c       Distance over which the image should be tapered down to the mean
c       intensity, or / to accept the default.  This taper will occur inside
c       of the drawn contour, so that the mean intensity is reached at the
c       position of the contour.
c       
c       David Mastronarde 12/4/98
c       
c       
      use fortmodel
      parameter (imsiz=4200)
      parameter (mxd=50)
      real*4 array(imsiz*imsiz),title(20),orig(3),delt(3)
      integer*4 nxyz(3),mxyz(3)
      equivalence (nx,nxyz(1)),(ny,nxyz(2)),(nz,nxyz(3))
      character*320 infile,outfile,ptfile
      real*4 xfix(5000),yfix(5000)
      logical readw_or_imod, typeonlist
      common /bigarr/array
      character*80 titlech
c       
      call setExitPrefix('ERROR: EDGEERASER - ')
      write(*,'(1x,a,$)')'Input image file: '
      read(*,'(a)')infile
      call imopen(1,infile,'old')
      call irdhdr(1,nxyz,mxyz,mode,dmin,dmax,dmean)
      if(nx*ny.gt.imsiz**2)call exitError('IMAGE TOO LARGE')
      call irtdel(1,delt)
      call irtorg(1,orig(1),orig(2),orig(3))
c       
      print *,'Enter output file name (Return to put modified'//
     &    ' sections back in input file)'
      read(*,'(a)')outfile
c       
      imfilout=1
      if(outfile.ne.' ')then
        imfilout=2
        call imopen(2,outfile,'new')
        call itrhdr(2,1)
      endif
c       
75    write(*,'(1x,a,$)')'Model file: '
      read(*,'(a)')ptfile
      if(.not.readw_or_imod(ptfile))then
        print *,'Bad model file, try again'
        go to 75
      endif

      do i=1,n_point
        do j=1,3
          p_coord(j,i)=(p_coord(j,i)+orig(j))/delt(j)
        enddo
      enddo
c       
      ntaper=10
      write(*,'(1x,a,i2,a,$)')'Distance over which to taper image '
     &    //'(/ for',ntaper,'): '
      read(5,*)ntaper
c       
c       start looping on sections; need to read regardless
c       
      do izsect=0,nz-1
        call irdsec(1,array,*99)
        nfix=0
c         
c         scan through points to see if any need fixing
c         
        do iobj=1,max_mod_obj
          if(npt_in_obj(iobj).gt.0)then
            ibase=ibase_obj(iobj)
            if(nint(p_coord(3,object(ibase+1))).eq.izsect)then
              nfix=1
              ninobj=npt_in_obj(iobj)
              xs=p_coord(1,object(ibase+1))
              xn=p_coord(1,object(ibase+ninobj))
              ys=p_coord(2,object(ibase+1))
              yn=p_coord(2,object(ibase+ninobj))
              dx=abs(xn-xs)
              dy=abs(yn-ys)
              xmid=(xs+xn)/2.
              ymid=(ys+yn)/2.
              if(dx.gt.dy.and.ymid.lt.ny/2)then
c                 
c                 case 1: wipe out below the contour
c                 
                do ip=1,ninobj
                  ipt=object(ibase+ip)
                  xfix(ip)=p_coord(1,ipt)+1.
                  yfix(ip)=p_coord(2,ipt)+1.
                enddo
                call sortfix(xfix,yfix,ninobj)
                sum=0.
                do ix=1,nx
                  iys=interpline(xfix,yfix,ninobj,ix)
                  colsum=0.
                  do iy=iys,ny
                    colsum=colsum+array(ix+(iy-1)*nx)
                  enddo
                  sum=sum+colsum/(ny+1-iys)
                enddo
                denmean=sum/nx
                do ix=1,nx
                  iys=interpline(xfix,yfix,ninobj,ix)
                  iytap=iys+ntaper
                  do iy=1,iys
                    array(ix+(iy-1)*nx)=denmean
                  enddo
                  do iy=iys+1,iytap
                    f=float(iy-iys)/(iytap+1-iys)
                    array(ix+(iy-1)*nx)=(1.-f)*denmean+
     &                  f*array(ix+(iy-1)*nx)
                  enddo
                enddo
              elseif(dx.gt.dy.and.ymid.gt.ny/2)then
c                 
c                 case 2: wipe out above
c                 
                do ip=1,ninobj
                  ipt=object(ibase+ip)
                  xfix(ip)=p_coord(1,ipt)+1.
                  yfix(ip)=p_coord(2,ipt)+1.
                enddo
                call sortfix(xfix,yfix,ninobj)
                sum=0.
                do ix=1,nx
                  iys=interpline(xfix,yfix,ninobj,ix)
                  colsum=0.
                  do iy=1,iys
                    colsum=colsum+array(ix+(iy-1)*nx)
                  enddo
                  sum=sum+colsum/iys
                enddo
                denmean=sum/nx
                do ix=1,nx
                  iys=interpline(xfix,yfix,ninobj,ix)
                  iytap=iys-ntaper
                  do iy=iys,ny
                    array(ix+(iy-1)*nx)=denmean
                  enddo
                  do iy=iytap,iys-1
                    f=float(iys-iy)/(iys+1-iytap)
                    array(ix+(iy-1)*nx)=(1.-f)*denmean+
     &                  f*array(ix+(iy-1)*nx)
                  enddo
                enddo
              elseif(xmid.lt.nx/2)then
c                 
c                 case 3: wipe out to left of the contour
c                 
                do ip=1,ninobj
                  ipt=object(ibase+ip)
                  xfix(ip)=p_coord(1,ipt)+1.
                  yfix(ip)=p_coord(2,ipt)+1.
                enddo
                call sortfix(yfix,xfix,ninobj)
                sum=0.
                do iy=1,ny
                  ixs=interpline(yfix,xfix,ninobj,iy)
                  colsum=0.
                  do ix=ixs,nx
                    colsum=colsum+array(ix+(iy-1)*nx)
                  enddo
                  sum=sum+colsum/(nx+1-ixs)
                enddo
                denmean=sum/ny
                do iy=1,ny
                  ixs=interpline(yfix,xfix,ninobj,iy)
                  ixtap=ixs+ntaper
                  do ix=1,ixs
                    array(ix+(iy-1)*nx)=denmean
                  enddo
                  do ix=ixs+1,ixtap
                    f=float(ix-ixs)/(ixtap+1-ixs)
                    array(ix+(iy-1)*nx)=(1.-f)*denmean+
     &                  f*array(ix+(iy-1)*nx)
                  enddo
                enddo
              else
c                 
c                 case 4: wipe out to right
c                 
                do ip=1,ninobj
                  ipt=object(ibase+ip)
                  xfix(ip)=p_coord(1,ipt)+1.-yextra
                  yfix(ip)=p_coord(2,ipt)+1.
                enddo
                call sortfix(yfix,xfix,ninobj)
                sum=0.
                do iy=1,ny
                  ixs=interpline(yfix,xfix,ninobj,iy)
                  colsum=0.
                  do ix=1,ixs
                    colsum=colsum+array(ix+(iy-1)*nx)
                  enddo
                  sum=sum+colsum/ixs
                enddo
                denmean=sum/ny
                do iy=1,ny
                  ixs=interpline(yfix,xfix,ninobj,iy)
                  ixtap=ixs-ntaper
                  do ix=ixs,nx
                    array(ix+(iy-1)*nx)=denmean
                  enddo
                  do ix=ixtap,ixs-1
                    f=float(ixs-ix)/(ixs+1-ixtap)
                    array(ix+(iy-1)*nx)=(1.-f)*denmean+
     &                  f*array(ix+(iy-1)*nx)
                  enddo
                enddo
              endif
            endif
          endif
        enddo
c         
        call iclden(array,nx,ny,1,nx,1,ny,dmint,dmaxt,dmeant)
        tmin=min(tmin,dmint)
        tmax=max(tmax,dmaxt)
        tsum=tsum+dmeant
c         
c         write out if any changes or if new output file
c         
        if(nfix.gt.0.or. imfilout.eq.2)then
          call imposn(imfilout,izsect,0)
          call iwrsec(imfilout,array)
        endif
      enddo
c       
      tmean=tsum/nz
c       
c       7/7/00 CER: remove the encodes
c       
c       encode(80,109,title)
      write(titlech,109)   
      read(titlech,'(20a4)')(title(kti),kti=1,20)
109   format('EDGEERASER: bad edge replaced with taper to mean')
      call iwrhdr(imfilout,title,1,tmin,tmax,tmean)
      call imclose(imfilout)
      call exit(0)
99    call exitError('Reading file')
      end


      function interpline(xfix,yfix,ninobj,ix)
      real*4 xfix(*),yfix(*)
      if(ix.lt.xfix(1))then
        interpline=yfix(1)+0.9
        return
      endif
      do i=2,ninobj
        if(ix.lt.xfix(i))then
          interpline=yfix(i-1)+(ix-xfix(i-1))*(yfix(i)-yfix(i-1))/
     &        (xfix(i)-xfix(i-1)) + 0.9
          return
        endif
      enddo
      interpline=yfix(ninobj)+0.9
      return
      end

      subroutine sortfix(x,y,n)
      real*4 x(*), y(*)
      if(x(1).gt.x(n))then
        do i=1,n/2
          ni=n+1-i
          tmp=x(i)
          x(i)=x(ni)
          x(ni)=tmp
          tmp=y(i)
          y(i)=y(ni)
          y(ni)=tmp
        enddo
      endif
c       
      do i=1,n-1
        do j=i+1,n
          if(x(i).gt.x(j))then
            tmp=x(i)
            x(i)=x(j)
            x(j)=tmp
c             
c             this is not perfect - it's better NOT to swap the Y vales
c             tmp=y(i)
c             y(i)=y(j)
c             y(j)=tmp
          endif
        enddo
      enddo
      return
      end
